src/tabs/ProgrammableREPL/index.tsx

```diff
@@ -0,0 +1,106 @@
+/**
+ * Tab for Source Academy Programmable REPL module
+ * @module programmable_repl
+ * @author Wang Zihan
+ */
+
+import React from 'react';
+import { DebuggerContext } from '../../typings/type_helpers';
+import { Button, ButtonGroup, Divider, NumericInput } from '@blueprintjs/core';
+import { IconNames } from '@blueprintjs/icons';
+import { ProgrammableRepl } from '../../bundles/programmable_repl/programmable_repl';
+import AceEditor from 'react-ace';
+import 'ace-builds/src-noconflict/mode-javascript';
+import 'ace-builds/src-noconflict/theme-twilight';
+import 'ace-builds/src-noconflict/ext-language_tools';
+
+
+type GUI_Data = {
+  programmableReplInstance: ProgrammableRepl;
+};
+
+class ProgrammableReplGUI extends React.Component<GUI_Data> {
+  public replInstance : ProgrammableRepl;
+  // private _editorInstance : any;
+  constructor(data: GUI_Data) {
+    super(data);
+    this.replInstance = data.programmableReplInstance;
+    this.replInstance.setTabReactComponentInstance(this);
+  }
+  public render() {
+    const outputDivs : JSX.Element[] = [];
+    const outputStringCount = this.replInstance.outputStrings.length;
+    for (let i = 0; i < outputStringCount; i++) {
+      outputDivs.push(<div style={{ color: this.replInstance.outputStrings[i].color }} dangerouslySetInnerHTML={ { __html: this.replInstance.outputStrings[i].content }} />);
+    }
+    return (
+      <div>
+        <Button
+          className="programmable-repl-button"
+          icon={IconNames.PLAY}
+          active={true}
+          onClick={() => this.replInstance.RunCode()}// Note: Here if I directly use "this.replInstance.RunCode" instead using this lambda function, the "this" reference will become undefined and lead to a runtime error when user clicks the "Run" button
+          text="Run"
+        />
+        <Button
+          className="programmable-repl-button"
+          icon={IconNames.FLOPPY_DISK}
+          active={true}
+          onClick={() => this.replInstance.saveEditorContent()}// Note: Here if I directly use "this.replInstance.RunCode" instead using this lambda function, the "this" reference will become undefined and lead to a runtime error when user clicks the "Run" button
+          text="Save"
+        />
+        <AceEditor
+          ref={ (e) => this.replInstance.SetEditorInstance(e?.editor)}
+          style= { {
+            width: '100%',
+            height: '375px',
+            ...(this.replInstance.customizedEditorProps.backgroundImageUrl != 'no-background-image' && {
+              backgroundImage: `url(${this.replInstance.customizedEditorProps.backgroundImageUrl})`,
+              backgroundColor: `rgba(20, 20, 20, ${this.replInstance.customizedEditorProps.backgroundColorAlpha})`,
+              backgroundSize: '100%',
+              backgroundRepeat: 'no-repeat',
+            }),
+          } }
+          mode="javascript" theme="twilight"

```

Comment:
Improvements in the future to consider? Maybe let users set the theme value for themselves? That would be quite cool imo.